### DISCUSSION TITLE
What are the backend we're thinking of?

### DISCUSSION THREAD

**Discussion by AbhinavMir** [NostrGit/NostrGit#73] [General]
I'm assuming the stack would be something like nostr, gitscm (handled by say, Go) hosting git on ipfs or self-hosted servers -> api exposed -> react renders the API.

**Comment by PeerRich** [comment]
ipfs or peer to peer torrent would be nice

i think issues, pull request descriptions, discussions etc should be normal nostr notes on relays

**Comment by gourcetools** [comment]
+1 peer to peer torrent

**Comment by AbhinavMir** [comment]
p2p torrent seems more permanent, IPFS has a recurring permanence issue currently, I'm not sure why. I think having mirrors would be helpful too, and latest image of torrent can be stored as a stack, for which we check the checksum of (and use a distance checksum) and do an ARIES-style rollback.

**Comment by PeerRich** [comment]
maybe we leave it to the user to choose? maybe we support multiple ways to store a repo

**Comment by davidabram** [comment]
One glaring issue with having repos exclusively over p2p torrent is that you can potentially lose  some history / commits. There would be possibility of some commits simply not be seeded anywhere.

I think we should allow users / orgs to self-host git servers also, to be more of a layer on top connecting all git servers vs a centralized forge.

Git over p2p torrent or nostrgit hosted git servers would help lower the barrier of entry, so an user or an org that wants to join doesn't need to self-host their git server to start using nostrgit.

**Comment by orthzar** [comment]
> torrents

[GitTorrent](https://github.com/cjb/GitTorrent) is one option. Here's the [blog post](https://blog.printf.net/articles/2015/05/29/announcing-gittorrent-a-decentralized-github/) describing it in detail. GitTorrent would require substantial modification to be usable via Nostr (e.g. replacing the code that talks to Bitcoin with code that talks to Nostr instead), but it at least provides a working means to store/retrieve git repos via torrents.

>maybe we leave it to the user to choose? maybe we support multiple ways to store a repo

The trade-off is that it would increase the complexity of Nostr clients, because each implementation would have to support each storage method. Also, the list of storage methods that a client must implement is something that would ideally be standardized in a NIP.

>I think we should allow users / orgs to self-host git servers also, to be more of a layer on top connecting all git servers vs a centralized forge.

The problem with self-hosted git servers is that they are all centralized. The point of the bounty is for the important aspects of Github (e.g. identity and storage) to be made decentralized. I suspect Dorsey would not award the bounty to any project that did not provide decentralization of git repo storage by default. I recommend not supporting centralized storage methods, because the time spent on that could be spent on other things (e.g. thinking about decentralized storage options, which gets the project closer to winning the bounty).

**Comment by cypherhoodlum** [comment]
I'm leaning towards trying out [GitTorrent](https://github.com/cjb/GitTorrent). We can always add support for self-hosted storage methods later if needed. GitTorrent could solve a deeper issue of decentralisation, as @orthzar mentioned.

**Comment by brycedev** [comment]
https://github.com/nip17/nips/blob/master/17.md

**Comment by cypherhoodlum** [comment]
Cool! :rocket:  was unaware of @melvincarvalho 's work.

**Comment by akhavr** [comment]
Git _is_ the decentralized storage by itself.

**Comment by PeerRich** [comment]
of course git is decentralised but we still need a clear web repo that people can push upstream

git torrent looks promising

**Reply by akhavr** [reply]
> of course git is decentralised but we still need a clear web repo that people can push upstream

Well, in https://github.com/akhavr/nostrya I'm going to go other route.
 
> git torrent looks promising

I'd suggest to read initial note by @fiatjaf why did he choose to drop ipfs and design nostr.  Hint: ipfs and torrent use similar underlying protocol to transfer big files.

**Reply by PeerRich** [reply]
can you link the post from @fiatjaf ?

**Reply by melvincarvalho** [reply]
A few things:

1. Article: https://fiatjaf.com/d5031e5b.html (I agree with this)
2. Gittorrent last commit 8 years ago: https://github.com/cjb/GitTorrent
3. Part of the design of nostr was inspired by torrent trackers similar to relays, trackers dont contain the data but relay messages from one client to another

**Reply by melvincarvalho** [reply]
Notes at the bottom of fiatjaf's article also good

**Reply by cypherhoodlum** [reply]
> Gittorrent last commit 8 years ago

Well that's pretty bad.

**Reply by cypherhoodlum** [reply]
fiatjaf's article made me lose hope with IPFS. I still fear forcing file hosting on nostr relay operators would push small operators out. Of course smaller relays could refuse to accept pushes, but is that the best solution?

**Comment by davidabram** [comment]
For the last few days I was trying to make backend work.

I have played with [GitTorrent](https://github.com/cjb/GitTorrent), [Nostream](https://github.com/Cameri/nostream) and [git server](https://hub.docker.com/r/jkarlos/git-server-docker/).

I am trying to solve how to store git commits and how to open PRs (send diff patches) in a distributed manner.

I haven't considered IPFS to store git commits. It highly possible I am straight wrong




These are the solution I have found that could work:



## Create a new NIP that supports git commits on the relay

This would be the most naive solution. It means that NostrGit would get all the git commits from the nostr relays.

### Pros:
-  from standpoint of NostrGit is the simplest, straight forward solution

### Cons:
- potentially enormous amount of data sent to relays
- potentially enormous amount of data transferred via websockets

I don't think this is feasible or that this is even a nostr use case. 

## Use BitTorrent protocol for hosting git commits and use nostr as DHT and to send diff patches

So [GitTorrent](https://github.com/cjb/GitTorrent) is really cool. I personally see it as an experiment that had potential. I don't think it's a finished product. GitTorrent can be taken as a blueprint.

### Pros:
- git commits are distributed
- low entry barrier for users -> git clone and seed it
- performant git clone performance via bittorrent

### Cons:
- if nobody seeds you can't git pull or git clone even tho "trackers" say there is a new commit / the repo exists
- No single source of truth unless every repo and fork has its own nostr pub/sec key and you trust the author of it
- it's really hard to fight against phishing / malware attacks -> you are pulling code from a torrent and running it :| (just see the number of malware packages currently on npm)

I would really like this solution work, but I my educated guess is that the end product would be almost akin to a shitty and slow git dapp then a working solution.

## Adopt NIP-17  [NIP-17](https://github.com/nip17/nips/blob/master/17.md) by ([melvincarvalho](https://github.com/melvincarvalho)) + [send diff patches via nostr](http://git.jb55.com/git-nostr-tools/file/README.txt.html)

Personally I think this is the most feasible approach. This would mean that git servers are self hosted / managed by users / orgs and issues, PRs and everything else is communicated via nostr.

### Pros:
- low entry barrier for users -> host your git server or create new repo on github
- NostrGit doesn't force users to store git commits in unorthodox way and doesn't force to user to drop their current git repos
- in my opinion the best starting place with potential to iterate on it (figure out decentralized storage etc.)
- low probability of runtime failure (either by missing or slow data transfer)

### Cons:
- git commits aren't stored in a decentralized manner, even tho git server are de facto decentralized.
- it's more of Github++ (most repos would still be hosted on popular furnaces)

I think this is the most feasible solution, this would make so nostr adopts a NIP that would help us tremendously, and it's a solution on which we can improve and iterate on without alienating any current furnace user. @melvincarvalho If I have missed something or said something incorrectly please let me know.

Quick sketch of how would it work:

![sketch](https://user-images.githubusercontent.com/7620347/223607927-5c2bdb4b-833c-4293-9fc6-cdbc2c42b7f6.png)

some-org has some-repo, and they want to use NostrGit to recieve Zaps, write Issues, get PRs from nostr community. some-org can choose which relays they want to send their git commit metadata (NIP-17). Any dev can fork this repo and create some changes. They open a PR by sending diff patches to any relay some-org is using. Now a PR is automatically opened on nostr.git/some-org/some-repo. 

## More out there solutions that I have dismissed because they aren't feasible at all:

NostrGit hosts git servers on a decentralized storage network and uses nostr for everything else
NostrGit creates a new low-level decentralized protocol that works with nostr protocol, where you store git commits

**Reply by PeerRich** [reply]
great research! i have my own security concerns with malware from torrents

sitting “on top” of github could be a good starting point and then increase decentralisation

as long as our project allows anyone to chose the git server i dont mind

github would be somewhat similar to a custodial wallet (greater convenience, less decentralized) whereas your own server = your own keys

maybe you can also chose to stream to multiple servers like mirrors

**Comment by fiatjaf** [comment]
I've written my thoughts on this, more or less, here: https://github.com/nostr-protocol/nips/pull/223#issuecomment-1456364051

I would like to learn what you think about that.

**Reply by PeerRich** [reply]
great to see you here.

getting an error: 
![image](https://user-images.githubusercontent.com/8019099/223693400-31011a22-bbfb-48ac-acc7-edacb61bb32b.png)

**Reply by fiatjaf** [reply]
The content of the note is copied on that GitHub comment.

**Reply by melvincarvalho** [reply]
> code should probably be hosted by standalone dedicated git servers -- and there could be centralized providers offering these services but they should interoperate seamlessly between themselves and with standalone personal servers

I liked this part.  I had this in mind too as a possible workflow.

**Comment by cypherhoodlum** [comment]
We [decided](https://github.com/NostrGit/NostrGit/discussions/95) to not use Discussions anymore. Let's continue the discussion at #115 :heart:
